### ISSUE TITLE
Событие ПослеУстановкиСтатусаДокумента

### ISSUE THREAD

**Issue by milanse** [diadoc/diadoc1cpro-docs#113]
При поступлении документа мне нужно провести анализ контента извлечь из него данные в допреквизиты элемента диадок_документы, возможно связать с существующим документом 1С. Как это правильно сделать ? 

Пытаюсь использовать событие ПослеУстановкиСтатусаДокумента но не понятно как данное событие работает. 
Вызывается ли оно при первом поступление документа от контрагента к нам на подпись и создании элемента справочника диадок_документы ? Есть подозрение что нет.

**Comment by DmitryModin** [comment]
Добрый день!
Уточните, пожалуйста, с какой версией модуля работаете? Лучше приложить техническую информацию из модуля.
Спасибо!

**Comment by milanse** [comment]
> Добрый день! Уточните, пожалуйста, с какой версией модуля работаете? Лучше приложить техническую информацию из модуля. Спасибо!

Техническая информация:

Версия модуля: 4.17.4.1922
Версия плагина: 1.4.3.195
Модуль типовой: Не удалось получить контрольную сумму файла
Расположение модуля: ВКонфигурации
Имя модуля: 
Конфигурация:  (1.2.2.71)
Модуль интеграции: 

Подключаемый модуль в конфигурации: ДД_подключаемыйМодуль
Тип платформы: Windows x86
Версия платформы: 8.3.17.2231
Версия ОС: Microsoft Windows 10 version 10.0  (Build 18363)
Вид клиента: Толстый клиент управляемые формы
Строка соединения: File="C:\1C\БАЗА";
Каталог временных файлов: C:\Users\sse\AppData\Local\Temp\

Настройки модуля:
Режим прокси: без прокси
Загружать исходящие документы: Нет
Загружать входящие документы: Да
Загружать документы только по сопоставленным контрагентам: Нет
Использовать сопоставление подразделений в документах на отправку: Нет
Разрешить фоновые операции: Нет
Сценарий сохранения документов: весь документооборот
Хранить документы на диске: Нет
Загружать документооборот целиком: Нет
Использовать типовые объекты 1С: Нет
Режим логирования: Нет
Режим отладки: Нет
Отключить контроль дублирования: Нет
Показать количество документов в пакете в разделе "Для отправки": Да
Удалять при переотправке предыдущий документ: Да
Создавать счета-фактуры: Да
Создавать передаточные документы: Нет
Отключить анализ для входящих документов: Нет
Указывать отсутствие основания УПД: Нет
Отправлять непроведенные документы: Нет
Записывать комментарий при отправке: Нет
Записывать комментарий при создании/сопоставлении: Нет
Использовать сервис "Контур.Маркировка": Нет

Настройки Стандарта:
Подключаемый модуль : отсутствует
Режим совместимости форматов 155 и 820: Нет

BoxId: f686d48a95424eb19f0901748097fdea@diadoc.ru
Отпечаток сертификата: 
Login для обмена: mail@mail.ru
UserId: 356c8400-730a-4241-89b2-fbe2214d55ce

Подключенные библиотеки:
 - Ядро вер. 1.1.1.1 (Макет)
 - ГенерацияXML вер. 4.11.1.135 (Макет)
 - ОсновнойМодуль вер. 4.17.4 (Макет)
 - ПарсерJSON вер. 1.0.0.2 (Макет)
 - АсинхронныеМетоды вер. 0.1.0.2 (Макет)
 - ВнешниеСервисы вер. 0.4.1.4 (Макет)
 - МетодыAPI вер. 0.1.0.10 (Макет)
 - ВызовыПМ вер. 0.4.1.19 (Макет)
 - ПечатныеФормы вер. 1.5.5.99 (Макет)
 - ПМ вер. 3 (Макет)
 - ХранениеДанных вер. 1.0.0.34 (Макет)
 - Плагин вер. 1.1.16 (Макет)
 - ЭмуляторТабДокумента вер. 0.1.0.2 (Макет)
 - УведомленияПользователей вер. 1.0.2.1 (Макет)
Пропустить обновление библиотек: Нет

**Comment by milanse** [comment]
@DmitryModin , я подскажу, в этой версии не работает, что, на мой взгляд странно. 

Вызывается из Лента_ОбновитьДанныеДокумента 

Вообще странно построена работа с входящими документами. Я думаю тут многие согласятся, что есть вещи, которые надо делать именно при поступлении нового документа - например маршрутизация, заполнение допреквизитов, заполнение реквизитов связанных документов 1С, может автоматическая их генерация, на все это имеем процедуру - после обработки ленты событий, без описания приемов работы с ней.

Может стоит сделать событие при поступлении пакета или как-то так?

**Comment by NikitaLysenkoA** [comment]
@milanse, для этих целей следует использовать событие "ПослеОбработкиЛентыСобытий" https://developer.kontur.ru/Docs/Diadoc_UM/func/pm/Posle_Obrabotki_Lenty_Sobytiy.html
Событие вызывается в том числе при первом поступление документа от контрагента и создании элемента справочника диадок_документы

Либо использовать анализ документов. Событие "ПровестиАнализДокумента" https://developer.kontur.ru/Docs/Diadoc_UM/func/pm/Provesti_Analiz_Dokumenta.html

Есть конкретный пример с кодом как и что пытаетесь заполнить?

**Comment by milanse** [comment]
@NikitaLysenkoA , спасибо, Я уже разобрался. Только параметры процедуры не описаны. Как понимать, что документ новый ?

Не понятно когда вызывается Провести анализ документа. У меня туда не попадает при синхронизации. Видимо должно работать по кнопке провести анализ  из формы?

**Comment by NikitaLysenkoA** [comment]
@milanse 
Да, работает по кнопке "Выполнить анализ" из формы списка. https://developer.kontur.ru/Docs/Diadoc_UM/proc/pm/Obrabotka_Vkhodyashchikh_Paketov.html?highlight=провестианализ#id10

Либо можно вызвать в ПМ один из методов:
ОсновнойМодуль.ЭДО_ПровестиАнализВходящегоДокумента(ДокументЭДО, ПакетЭДО)
ОсновнойМодуль.ЭДО_ПровестиАнализВходящегоПакета(ПакетЭДО)

Узнать что пакет пришел впервые стандартными средствами нельзя, тут нужно реализовывать свою логику. Например писать в пакет дополнительный реквизит в табличную часть "ДополнительныеРеквизиты"

**Comment by milanse** [comment]
@NikitaLysenkoA, Я обо всем этом написал выше. Если дергать провести анализ пакеты программно при синхронизации, понимание того что пакет только что поступил просто необходимо, иначе будет дёргаться при каждом изменении документа / пакета.

**Comment by NikitaLysenkoA** [comment]
@milanse в таком случае нужно реализовывать свою логику.

**Comment by milanse** [comment]
@NikitaLysenkoA , ну тут же принимаются идеи на доработку.
Хотелось бы иметь обработчик события при поступлении пакета, при поступлении документа. 

В событии выполнить анализ пакета и документа иметь признак, что он новый.

Событие при смене статуса документа вызывать и при его первой загрузке. В событие прокинуть старый статус, тем более что в кишках все это  вычисляется. 

Вы мне предлагаете писать логику снаружи, а между тем все это есть у вас внутри моделей.

**Comment by DmitryModin** [comment]
Добрый день!
Пожелания конечно же принимаются. 
Ваше пожелание, на обработчики событий при получении пакета\документа и признак, что документ новый, зафиксировано.
Спасибо!

**Comment by acharushkin** [comment]
Здравствуйте!

Мы архивируем проект `diadoc1cpro-docs`.

Актуальная документация по модулю Диадока для 1С находится здесь:

- [Для разработчиков](https://developer.kontur.ru/docs/diadoc_um/index.html)
- [Для пользователей](https://support.kontur.ru/diadoc-1s8x)

Наши контакты:

- [Клиентская поддержка Контур.Диадок](https://kontur.ru/support#diadoc) - для консультации по ошибкам в работе модуля
- [Заявка на интеграцию](https://www.diadoc.ru/order1c#order) - для подключения интеграции и связанных вопросов. Например, если нужен сертификат, узнать тарифы и стоимость подключения

С уважением, команда разработки Диадока
